### ISSUE TITLE
Learning Non-negative Parameters

### ISSUE THREAD

**Issue by Taha-Bahadori** [PyTorchLightning/pytorch-lightning#2382]
How can I enforce non-negativity of weights in pytorch-lightning?

In PyTorch, I use `parameter.clamp(min=0)` after every optimization update to perform the projected gradient descent. Where should I do this in pytorch-lightning?

**Comment by rohitgr7** [comment]
You can override `optimizer_step` in your `LightningModule` class. Call `optimizer.step()` and then `self.parameters.clamp(min=0.)`.
https://github.com/PyTorchLightning/pytorch-lightning/blob/2f739f5977640bb8580b82a11f322f81f1b90d09/pytorch_lightning/core/lightning.py#L1128-L1143

**Comment by Taha-Bahadori** [comment]
@rohitgr7 Thank you.

**Comment by akihironitta** [comment]
@arjunsinghrathore I think you need to call the method on `torch.Tensor` instance of your interest but not on `function`, e.g.

```python
your_tensor.clamp(...)
```

See usage here: https://pytorch.org/docs/stable/generated/torch.Tensor.clamp.html#torch.Tensor.clamp

**Comment by arjunsinghrathore** [comment]
> @arjunsinghrathore I think you need to call the method on `torch.Tensor` instance of your interest but not on `function`, e.g.
> 
> ```python
> your_tensor.clamp(...)
> ```
> 
> See usage here: https://pytorch.org/docs/stable/generated/torch.Tensor.clamp.html#torch.Tensor.clamp

Here I am trying to clamp the weights of the whole model and not a single tensor. Based on the above discussion `self.parameters.clamp(min=0.)` should work but I'm getting an error. If you know some other way for clamping the weights in pytorch lightning then please do share?

**Comment by akihironitta** [comment]
> Based on the above discussion self.parameters.clamp(min=0.) should work 

@arjunsinghrathore I believe it's wrong then because `self.parameters` within `LightningModule` is reserved by `torch.nn.Module` ([docs](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.parameters)), and it just returns a generator.

If you want to `clamp` the whole parameters in your `LightningModule`, I think you'll need to do the following:

```python
for param in self.parameters():
    param.clamp(...)
```

**Comment by jermwatt** [comment]
Here is a complete example

```python
    def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_closure):
        # step
        optimizer.step(closure=optimizer_closure)

        # project
        for param in self.parameters():
            param.data = param.data.clamp(min=0.)
```

Note that you need to update ```param.data``` which is the ```<class 'torch.Tensor'>``` object in each layer.  The ```param``` object is of type ```<class 'torch.nn.parameter.Parameter'>```.
